text/0474-event-bridge-scheduler-l2.md

```diff
@@ -0,0 +1,481 @@
+> section of your RFC. It's a powerful thought exercise which will challenge you
+> to truly think about this feature from a user's point of view.
+>
+> Choose *one or more* of the options below:
+>
+> * **CHANGELOG**: Write the changelog entry for this feature in conventional
+>   form (e.g. `feat(eks): cluster tags`). If this change includes a breaking
+>   change, include a `BREAKING CHANGE` clause with information on how to
+>   migrate. If migration is complicated, refer to a fictional GitHub issue and
+>   add its contents here.
+>
+> * **README**: If this is a new feature, write the README section which
+>   describes this new feature. It should describe the feature and walk users
+>   through usage examples and description of the various options and behavior.
+>
+> * **PRESS RELEASE**: If this is a major feature (~6 months of work), write the
+>   press release which announces this feature. The press release is a single
+>   page that includes 7 paragraphs: (1) summary, (2) problem, (3) solution, (4)
+>   leader quote, (5) user experience, (6) customer testimonial and (7) one
+>   sentence call to action.
+
+**CHANGELOG**:
+
+`feat(events-scheduler): Event Bridge Scheduler L2 constructs`
+
+ **README**: 
+
+ See below
+
+ # Amazon EventBridge Scheduler Contruct Library
+
+[Amazon EventBridge Scheduler](https://aws.amazon.com/blogs/compute/introducing-amazon-eventbridge-scheduler/) is a feature from Amazon EventBridge that allows you to create, run, and manage scheduled tasks at scale. With EventBridge Scheduler, you can schedule one-time or recurrently tens of millions of tasks across many AWS services without provisioning or managing underlying infrastructure.
+
+1. **Schedule**: A schedule is the main resource you create, configure, and manage using Amazon EventBridge Scheduler. Every schedule has a schedule expression that determines when, and with what frequency, the schedule runs. EventBridge Scheduler supports three types of schedules: rate, cron, and one-time schedules. When you create a schedule, you configure a target for the schedule to invoke. 
+2. **Targets**: A target is an API operation that EventBridge Scheduler calls on your behalf every time your schedule runs.  Targets can include Amazon EC2 instances, AWS Lambda functions, Kinesis streams, Amazon ECS tasks, Step Functions state machines, Amazon SNS topics, Amazon SQS queues, Amazon CloudWatch LogGroups, and built-in targets. A target receives events in JSON format.
+3. **Schedule Group**: Scheduling groups help you organize your schedules. Scheduling groups support tags that you can use for cost allocation, access control, and resource organization. When creating a new schedule, you can add it to a scheduling group.
+
+This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project. It allows you to define Event Bridge Schedules.
+
+
+## Defining a schedule 
+
+```ts
+
+const target = new targets.LambdaInvoke(props.func, {
```

Comment:
and if so, `events` already has a concept of a schedule. Can you explain what this resource achieves that the schedule class in `events` does not?

Reply:
> this is coming from the existing events-targets module right?

No it's a separate module `targets` for schedule targets. I do not think that we should attempt to reuse the same module for EB Rules Targets and EB Schedule Targets. I think as in CFN those targets are different CFN declarations ([Rule Target CFN](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-scheduler-schedule-target.html), [EB Target CFN](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-events-rule-target.html)), attempting to somewhat reuse these classes in L2 constructs will create more problems for edge-cases when behaviour between targets is different. 

Curious to know what do you think should be the reasons for reusing module targets? I would guess, that maybe in customer code it would be easier to use the same targets for EB rules and EB schedules, right?

Reply:
> Can you explain what this resource achieves that the schedule class in events does not?

In EB Rules a `Schedule` is just a schedule expression – a string that defines on what schedule the rule should be invoked. 

In EB Scheduler a `Schedule` is a separate CFN resource. It has much more properties than just a schedule expression (https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-scheduler-schedule.html). For example, `Schedule` supports timezone, groups, encryption, it can be disabled. `Schedule` in Scheduler plays similar role to  `Rule`: it is a central concept of the package that connects event input and event target.

Also, in the prototype implementation of L2 Scheduler constructs, we have a class `ScheduleExpression` that inherits `eb.Schedule`. We have added a method `at(..)` that allows to create a one-time schedule from a date. See https://github.com/filletofish/cdk-eb-scheduler/blob/main/lib/aws-scheduler/lib/expression.ts#L10